semantic/annotate.go

```diff
@@ -0,0 +1,67 @@
+package semantic
+
+import (
+	"strconv"
+)
+
+type typeVar struct {
+	prefix string
+	suffix int
+}
+
+// TypeAnnotationVisitor visits the nodes in a semantic
+// graph and assigns all sub-expressions type variables.
+type TypeAnnotationVisitor struct {
+	// tenv represents the type environment of a program.
+	// It maps a semantic node to a unique type variable.
+	tenv map[Node]Type
+	// the next type variable to be assigned. The first
+	// type variable is t0. Then t1, t2, t3 and so on.
+	next *typeVar
+}
+
+// Visit assigns a new type variable to a semantic node
+func (v *TypeAnnotationVisitor) Visit(node Node) Visitor {
+	if n, ok := node.(Expression); ok {
+		v.tenv[n] = v.newTypeVar()
+	}
+	switch n := node.(type) {
+	case *CallExpression:
+		if n.pipe != nil {
+			v.tenv[n.pipe] = v.newTypeVar()
+		}
+	case *Identifier:
+		v.tenv[n] = v.newTypeVar()
+	case *Property:
+		v.tenv[n] = v.newTypeVar()
+	case *FunctionParam:
+		v.tenv[n] = v.newTypeVar()
+	}
+	return v
+}
+
+// Done is required to implement the Visitor Interface
+func (v *TypeAnnotationVisitor) Done() {}
+
+// NewTypeAnnotationVisitor returns a new TypeAnnotationVisitor
+func NewTypeAnnotationVisitor() *TypeAnnotationVisitor {
+	return &TypeAnnotationVisitor{
+		tenv: make(map[Node]Type, 64),
+		next: &typeVar{
+			prefix: "t",
+			suffix: 0,
+		},
+	}
+}
+
+// newTypeVar returns the next type variable to be assigned
+func (v *TypeAnnotationVisitor) newTypeVar() Type {
```

Comment:
A TypeVar is not a Type. I think we need a new type called `typeVar`. The existing `typeVar` struct should probably be renamed to `typeVarGenerator` since its purpose it to create a new type var.

Then the `typeVar` name is free to become its own type, probably something like this:

```
type typeVarGenerator struct {
	prefix string
	suffix int
}

func (g *typeVarGenerator) Generate() typeVar {
   //...
}
type typeVar struct {
    name string
   // maybe need more here once we move on to unification...
}
```

And then the `tenv map[Node]Type` would become `tenv map[Node]typeVar`.

Thoughts?

Reply:
@nathanielc So I'm currently thinking through this. Are you saying remove `TypeVar` and instead create a `typeVar` that implements the `Type` interface (so it is still considered a type) but when you call `Kind()` on a `typeVar` if it doesn't know its concrete type it will just return `nil`?

I agree it doesn't make sense to have `TypeVar` be a primitive type as it is only going to be used internally during type inference. My reason for doing it originally was with the substitution phase in mind. During substitution we will be building up our types. In this phase type variables might be assigned to concrete types like int, bool, etc. But they might also be assigned to not yet complete types (types that are composed of type variables). For example:

    t0 = f(t1, t2) -> int

What is `f(t1, t2) -> int` here? It's not a type variable, but it's also not a function type because its inputs are two type variables. If `typeVar` implements `Type`, then both sides of the above equation are types, and we can then perform substitutions on types.

Or does it not make sense for `typeVar` to implement `Type`?

Reply:
yeah, I do not think `typeVar` should implement Type. It should have its own methods for reporting what the concrete type is. Maybe just a `Type() Type` method on the `typeVar` struct? So instead of `typeVar` being a type it can tell you its Type.